nuxeo-core/nuxeo-core-storage-dbs/src/main/java/org/nuxeo/ecm/core/storage/dbs/DBSTransactionState.java

```diff
@@ -402,31 +422,156 @@ public void updateAncestors(String id, int ndel, Object[] ancestorIds) {
         }
     }
 
+    protected int getReadAclsAsyncThreshold() {
+        boolean enabled = Boolean.parseBoolean(
+                Framework.getProperty(READ_ACL_ASYNC_ENABLED_PROPERTY, READ_ACL_ASYNC_ENABLED_DEFAULT));
+        if (enabled) {
+            return Integer.parseInt(
+                    Framework.getProperty(READ_ACL_ASYNC_THRESHOLD_PROPERTY, READ_ACL_ASYNC_THRESHOLD_DEFAULT));
+        } else {
+            return 0;
+        }
+    }
+
     /**
      * Updates the Read ACLs recursively on a document.
      */
     public void updateTreeReadAcls(String id) {
         // versions too XXX TODO
-        Set<String> ids = getSubTree(id, null, null);
-        ids.add(id);
-        ids.forEach(this::updateDocumentReadAcls);
+
+        save(); // flush everything to the database
+
+        // update the doc itself
+        updateDocumentReadAcls(id);
+
+        // check if we have a small enough number of descendants that we can process them synchronously
+        int limit = getReadAclsAsyncThreshold();
+        Set<String> ids = getSubTree(id, null, null, limit);
```

Comment:
we may need to check that all getSubTree impl are efficient on large volume

Reply:
I updated all implementations to try to be efficient in that regard. I agree that we'll know only when doing actual tests on huge volumes.